### PR TITLE
Integrating OSM data import

### LINKED ISSUE #1: Include OSM Import in egon-data
The OSM import including the corresponding scripts and stylefile needs to be included in the egon-data repository. We would like to keep the structure of the tables as it was in open_eGo but additional tags might trigger additional columns. 
Furthermore metadata and rtd documentation needs to be updated.

### PR DESCRIPTION
**Pull request by gplssm**
Fixes #1

### CODE REVIEW

**Comment by gplssm** [comment]
OSM data import scripts from open_eGo are now integrated by this PR.

@gnn I would ask you for a review, but you don't have to review resulting data. It is more or less the exact same way of importing OSM data as in open_ego. Requested changes can still added later.

The are some aspects that need to be discussed

- Shall we stick to specify the OSM data set of Bremen in [`data_sets.yml`](https://github.com/openego/eGon-data/blob/features/%231-osm-data-import/src/egon/data/data_sets.yml#L5) until we have the feature of running the workflow on a test region?
- Some helper functions are introduced and right now located in [utils.py](https://github.com/openego/eGon-data/blob/features/%231-osm-data-import/src/egon/data/utils.py). These might be interesting for others. Does it make sense to keep them there? Shall we change some of them?
- I used bitshift operators to specify the order of execution in `pipeline.py`. Explanation in commit message 
2276b3a. This is contradicting to your proposal. Please let's find one way to go to suggest it for the others
- Would it be ok to introduce not entirely tested code and write more tests later? We could already file an issue for that and assign it to me. The reason is that I would like to merge this PR very soon
- When I created this branch, I branched from `features/#18-start-postgres-docker-container-via-dag-task`. Meanwhile I merged dev into this one and the branch is up-to-date. Can I simply change the base branch of this PR?

I'll fix the failing tests/checks meanwhile. You can already start to review the other things above.

**Comment by gnn** [comment]
Since it's kinda hard to put reviews on the exactly appropriate places, I'll just answer your previous post and put additional comments at the end. Beware, this might get long.
 
> * Shall we stick to specify the OSM data set of Bremen in [`data_sets.yml`](https://github.com/openego/eGon-data/blob/features/%231-osm-data-import/src/egon/data/data_sets.yml#L5) until we have the feature of running the workflow on a test region?

Sure. Having something that works is better than not having something that *might* be better.


> * Some helper functions are introduced and right now located in [utils.py](https://github.com/openego/eGon-data/blob/features/%231-osm-data-import/src/egon/data/utils.py). These might be interesting for others. Does it make sense to keep them there? Shall we change some of them?

I'm incredibly glad that you ask. That gives me the opportunity to start with a rather unimportant remark: `utils` is an abbreviation. There's no harm in just calling it `utilities`, thus not setting a precedence for using abbreviations. It's not a problem in this case, but abbreviations can become very harmful in other contexts. Sorry for nit-picking.
Now that this is out of the way: having a `utilities` module is hinting at something fishy anyway. Regardless of whether you call it `utilities`, `tools`, `miscellaneous` (aka. `misc` (see, abbreviations again)) or something similar, it's all the same concept called a ["kitchen sink"][0] module. It's something without a specific purpose which can hold everything for which we can't find a more suitable place. In the long run, most projects end up having something like this. But because it can hold nearly everything, it's also dangerous, because it can have the effect of things just ending up in there without giving a proper place for it enough consideration. Your `utils` module contains a few good examples:

 - `egon_data_db_credentials`: We are already in the `egon.data` package, so there's no need to repeat that information here. Strip it of and you end up with `db_credentials`. Consider that underscores are slightly evil and should be avoided if possible and you can see that there might be a case for a `db` module with a `credentials` function. Doing so, you end up with a `egon.data.db.credentials` function, which looks a lot nicer than `egon.data.utils.egon_data_db_credentials`, at least IMHO.
You might go a step further and avoid the `db` abbreviation, spelling it out to `database`, but since `db` is such a common abbreviation, one could also make an exception here. Oh the lure of convenient inconsistencies. :)
- Now that we have a `database` module we can actually move two more functions from the `utils` module: `execute_sql` and `submit_comment` deal exclusively with database related stuff, so it's more suitable to place them into that module.
- `data_set_configuration`: That one's a weird one. I think it would be a good idea to use a proper configuration file for `egon.data`, where `dataset` is it's own section. That way we would have a `egon.data.configuration` module dealing with all configuration. The function can then be renamed to `egon.data.configuration.datasets` and we can later try to refactor the configuration to encompass more than just datasets. Also, notice that I removed the underscore. This is for various reasons:
    - "dataset" is a technical term, so you can use it.
    - It's not always a set, which is why `data_set` is misleading.
    - When I saw the name `data_set_configuration` I didn't know what it was about: configuring datasets or setting the configuration via data. It's better to avoid such confusion.
- Doing all of this, the `utils` module is actually gone.


> * I used bitshift operators to specify the order of execution in `pipeline.py`. Explanation in commit message
>   [2276b3a](https://github.com/openego/eGon-data/commit/2276b3aafdd92d267737e94eb88d5a97c77930a5). This is contradicting to your proposal. Please let's find one way to go to suggest it for the others

I'd suggest that both ways are OK. I simply didn't talk about the operators, because they can be confusing. It's the same symbol(s) doing totally different things depending on the context. And since Python isn't typed, figuring that context out is pretty hard. Simply put, seeing `a >> b` one doesn't really know what's happening, whereas `a.set_downstream(b)` is only ambiguous if there's more than one `set_downstream` method floating around. That being said, you solved it the correct way, by explaining the ambiguous code in the commit message. :+1: So if somebody is confused about what the code does, they can always do a `git blame` and look at the commit message of the confusing line and all will be clear. You didn't have to mention me and my talk though, because eventually somebody will read the message who doesn't know me and didn't listen to the talk, and these lines won't make sense to him. Simply explaining the code and linking to documentation would be enough. Last but not least, I think the operators are OK, because they keep the code short while still being readable. On the other hand, more complex pipelines have to be set up with functions anyway, so I wouldn't rule them out either.


> * Would it be ok to introduce not entirely tested code and write more tests later? We could already file an issue for that and assign it to me. The reason is that I would like to merge this PR very soon

Sure. It's our code and we're free to decide how much we sacrifice quality assurance for progress speed. ;) Also, that's why the badges are colored, so while we're below par, they constantly remind us of the necessity for improvement. :D


> * When I created this branch, I branched from `features/#18-start-postgres-docker-container-via-dag-task`. Meanwhile I merged dev into this one and the branch is up-to-date. Can I simply change the base branch of this PR?

Please do. Also, when you open a pull request, you can choose where to merge. It's generally a good idea to merge feature branches into dev and only merge feature branches into other feature branches if you're responsible for both and you know that the former was actually a sub-feature of the latter.


> I'll fix the failing tests/checks meanwhile. You can already start to review the other things above.

Oh, and I will. :) Here are the other things, that I stumbled upon.

I saw that you had to reformat your code a few times to make the style checkers happy. Sorry for not writing this anywhere, but in terms of code style, there's a tool for that. It's called `black`, you can install it via `pip install black`, it's configured via "setup.cfg" and it formats the files for you. I heavily recommend everybody to use it, to the point where I'm contemplating setting up a pre-commit hook that rejects commits which aren't properly formatted. Concerning the order of imports, here are the simple rules:

  - There are three sections, separated by a blank line: standard library imports, third party imports and imports from the local repository.
  - Within these sections, imports are sorted alphabetically. This is why `from` imports go before `import` imports.

I don't think you can get much simpler. Also, `isort` can actually do this automatically and it's configured via `setup.cfg`, too.
I know, these things need to be written down in the "CONTRIBUTING.rst" guidelines and I'll do it, when I get around to it. :)

I saw that you rolled back the commit renaming `data_import` to `import`, probably because import is a highly reserved word which can't be used for most things in Python except, well, imports. :) That's why I wanted to rename the module to `imports`. If you feel that this is confusing because you'd expect a collection of Python imports in the module, here are a few other candidates: "sources", "external", "foreign". On the other hand, this confusion can also be mitigated by a simple module docstring.

The last point has already been very nit-picky, but here are a few more very nit-picky things. I apologize in advance.

Phrasing "PEP8 E501 fixes" as "Fix PEP8 violations leading to E501 from flake8" is more consistent with [commit.style][1].

Similarly for commit 0130b74, the prefix "WIP: " is usually used for stashes not leaving your private repository. Something along the line of "Start integrating the OSM data import" with an in depth explanation of how such a big file suddenly springs into existence would have been nice. I can only assume, that you copied the file from the previous dataprocessing codebase. ;) Is that correct?

I think that's all I found. Let me conclude with the observation, that I could only nit-pick at two commits out of forty, so you're actually probably more disciplined than I am.

[0]: https://www.adam-bien.com/roller/abien/entry/util_packages_are_evil
[1]: https://commit.style/

**Comment by gplssm** [comment]
Todos for Guido

- [ ] Warning in docs that only Bremen is imported now
- [ ] rename `utils.py`to `db.py` and do other renaming proposals by Günni
- [ ] Rename `data_import/` to `imports/` or `importing`
- [ ] Create a module `egon.data.config.py` and put function `datasets()` in there (which is the `data_set_configuration()` but renamed). Returns information about all datasets
- [ ] Add issue for missing tests related to OSM import
- [ ] Apply black on code and if necessary isort again

**Comment by gplssm** [comment]
Two things regarding checks remain which I cannot resolve

- Codacy has 4 remaining complaints which I cannot resolve
- Travis fails only on macOS because it can't find Pthon 3.8, see `ERROR: InterpreterNotFound: python3.8`

I would be ok with merging, but then these failing checks will bother the next person. @gnn could you somehow fix these?

**Comment by gnn** [comment]
> * Codacy complains about "No blank lines allowed after function docstring (found 1) (D202)" [for example](https://app.codacy.com/gh/openego/eGon-data/file/50036355571/issues/source?bid=20606893&fileBranchId=20688897#l9)
> * Codacy has 4 other remaining complaints which I cannot resolve

Hmm. I need to figure out, what these complaints are about. At least the one about the multiline docstrings is not PEP257 compliant and I need to figure out how to disable it. The same probably goes for the others, but I need to check those first. But I wouldn't worry too much about the codacy checks. These are code quality checks and just with the coverage, it's OK if this is reduced for a while and then raised in later commits.


> * Travis fails only on macOS because it can't find Pthon 3.8, see `ERROR: InterpreterNotFound: python3.8`

Yeah, that's because the macOS build recipe now installs Python 3.9. I'll mark this as allowed to fail until I figure out how configure the build to use a different Python version.

**Comment by gplssm** [comment]
I agree with you! Since code quality isn't binary, we can accept some limitations in quality (for some time).

As I'm still looking forward to finally merge this PR, I would ask you to do just changes which do not take too long and then approve this PR. If you can move some things to later/elsewhere, I would be happy!

**Comment by gnn** [comment]
Sorry. I tried to do approve this fast, but while looking at the code I found some things and noted them down. Most of them don't require immediate action, but I at least wanted to put remarks in appropriate places, maybe discuss some things and open issues before approving the branch. I should get around to finishing this on Wednesday though.

**Review by gnn — COMMENTED** [review]
Just a few notes. None of this has to be done immediately, if at all. If you concur with some of the suggestions, we can simply open issues as TODOs for later.
I'm not done though. Will post more comments later.
Sorry for that.

File: src/egon/data/db.py
Comment (gnn) [inline src/egon/data/db.py:81]: As this is just a select statement, which doesn't really do anything, maybe you can put a short comment here, or on the executing line, explaining why this is done?

Comment (gplssm) [reply src/egon/data/db.py:81]: This simply queries the JSON string and casts it into JSON. It fails on invalid JSON. I guessed it was introduced in order to verify validity of the JSON string for correct displaying on OEP.

File: src/egon/data/db.py
Comment (gplssm) [reply src/egon/data/db.py:67]: Dunno :thinking:

Comment (gnn) [reply src/egon/data/db.py:67]: I was asking because it looks like that's the case, but I'm not too sure, too. If you don't know either, it might be a good candidate for a test, i.e. try out the behaviour first and then pretend it's intentional by writing a test which checks it. Fortunately, Travis [supports docker](https://docs.travis-ci.com/user/docker/).

File: src/egon/data/db.py
Comment (gplssm) [reply src/egon/data/db.py:72]: I ask @christian-rli for a location to link at that describes metadata string used with data uploaded to OEP. I'll add a link when  I have an answer.

Comment (christian-rli) [reply src/egon/data/db.py:72]: Description of individual keys and values: https://github.com/OpenEnergyPlatform/oemetadata/blob/develop/metadata/v140/metadata_key_description.md
Example string: https://github.com/OpenEnergyPlatform/oemetadata/blob/develop/metadata/v140/example.json
Schema: https://github.com/OpenEnergyPlatform/oemetadata/blob/develop/metadata/v140/schema.json

Comment (christian-rli) [reply src/egon/data/db.py:72]: Your choice :)

**Review by gnn — COMMENTED** [review]
These are the last comments. Sorry for taking so long.

File: src/egon/data/processing/openstreetmap.py
Comment (gnn) [inline src/egon/data/processing/openstreetmap.py:5]: This function should probably be split up into smaller functions. The fact that there are sections with explanatory comments are a good indicator that this should actually be dedicated functions with the comments as their docstrings.

File: src/egon/data/processing/openstreetmap.py
Comment (gnn) [inline src/egon/data/processing/openstreetmap.py:40]: Theres a name for the concept behind those changes. Updating the table schema like this is called a "migration". Theres a nice migration framework called "Alembic", built by the guy behind SQLAlchemy. It would probably be nice to figure out how to use that to encapsulate these changes. That's a TODO for myself though.

**Review by gnn — COMMENTED** [review]
Ok, I was lying. I do have to more comments. The good news is though, that the only thing that I would like to settle before merging this are the changes to the configuration file format. If this gets accepted and implemented, or rejected, this is good to go.
For all the other comments, issues can be made in order to work on them later. I also have a bunch of other code "improvements" planned, which I'll start incorporating on a separate branch, once this is merged.

File: src/egon/data/db.py
Comment (gplssm) [reply src/egon/data/db.py:77]: Thank your @christian-rli !

File: src/egon/data/processing/openstreetmap.py
Comment (gplssm) [reply src/egon/data/processing/openstreetmap.py:40]: I would love to learn more about migrations. Theoretically, I know what it does, but I've never used it. For now, I leave it on your TODO list ;-)

File: src/egon/data/db.py
Comment (gplssm) [reply src/egon/data/db.py:67]: I moved it to an issue: https://github.com/openego/eGon-data/issues/51

File: src/egon/data/datasets.yml
Comment (gplssm) [reply src/egon/data/datasets.yml]: I did more or less like you suggested (fa8cab24b7897a4838ff9b6fa347a6d4d14f3910). It is still inconsistent at some place, but I won't be too strict here.

File: src/egon/data/importing/openstreetmap/import_osm.py
Comment (gplssm) [reply src/egon/data/importing/openstreetmap/import_osm.py]: Done.

Actually, I cannot answer you question properly. I was more or less transferring the code from open_ego without understanding every detail.

File: src/egon/data/importing/openstreetmap/import_osm.py
Comment (gplssm) [reply src/egon/data/importing/openstreetmap/import_osm.py]: I had issue with downloading the file to CWD.

File: src/egon/data/importing/openstreetmap/import_osm.py
Comment (gplssm) [reply src/egon/data/importing/openstreetmap/import_osm.py]: Some software rely on naming it `geom` (maybe PostGIS). A least it is so super common that there is no need to change it.

File: src/egon/data/processing/openstreetmap.py
Comment (gplssm) [reply src/egon/data/processing/openstreetmap.py:5]: I would do that on-demand. If you prefer to file an issue for that already, feel free. Otherwise, please click "resolve".

File: src/egon/data/importing/openstreetmap/import_osm.py
Comment (gplssm) [reply src/egon/data/importing/openstreetmap/import_osm.py]: The SQL code is taken from old open_ego dataprocessing. The reason for doing this was conventions, I guess. We could actually remove it. But I makes easier to migrate other code from old open_ego DP which uses OSM data.

File: src/egon/data/datasets.yml
Comment (gplssm) [reply src/egon/data/datasets.yml]: See #53

**Comment by gplssm** [comment]
Thanks for all the comments and suggestions. I went through all of them and either implemented it, filed an issue or simply explained something as reply to your comment. All "uncritical" comments are marked as resolved. Only one remains open.
Could you please quickly check that you're happy with the PR and approve it if possible?

I noticed that Codacy has newly introduced import errors which I do not understand because the code runs smoothly.

I would like to merge today until 6 pm. Do think it is possible to receive an approval until then?

File: src/egon/data/datasets.yml
Comment (gnn) [reply src/egon/data/datasets.yml]: You're not, but I am. JK, I created issue #54 to remind me to have a look at it. Should be fine the way it is and can be merged.

File: src/egon/data/importing/openstreetmap/import_osm.py
Comment (gnn) [reply src/egon/data/importing/openstreetmap/import_osm.py]: Note to self: marked as resolved because we're using `osm_config["original_data"]["target"]["path"]` now, which can be a path (duh) which makes `os.path.basename` important again.

File: src/egon/data/importing/openstreetmap/import_osm.py
Comment (gnn) [reply src/egon/data/importing/openstreetmap/import_osm.py]: Ah, interesting. I opened issue #49 in order to look into this. Right now we're downloading data into the package installation location, which we might not always have write access to. Also, while it looks like here's the option to configure a target "path", we always prepend the package directory which might lead to unexpected results.
So this is something that needs fixing in the long run. But that's what the issue is for, so right now, things are probably OK enough.

File: src/egon/data/processing/openstreetmap.py
Comment (gnn) [reply src/egon/data/processing/openstreetmap.py:5]: It's related to the next conversation anyway, so I submitted it as #55.

File: src/egon/data/processing/openstreetmap.py
Comment (gnn) [reply src/egon/data/processing/openstreetmap.py]: Submitted as #55. See also the conversation above.

**Review by gnn — APPROVED** [review]
Go ahead and merge. I added notes and issues to everything I want to preserve for later, so the changes are fine. Also, don't worry about the Codacy complaints. Those import errors happen because they test a non-installed version of the codebase, so obviously every import of stuff that gets installed with the package will fail. I'll disable those checks, because import errors should be caught by our tests, not static code analysis.
Last but not least, please take the time to change the merge message to something readable like e.g. 

    Merge OSM import into dev

    This fixes #1.

or something similar.

Thanks and have fun.

**Pull request merged**
